syntax = "proto3";

package qkd;

// Service que Alice offre à Bob
service CascadeService {
  // Bob demande les parités de plusieurs blocs
  rpc AskParities (ParityRequest) returns (ParityResponse);
  
  // Bob signale le début d'une nouvelle session Cascade
  rpc StartReconciliation (StartRequest) returns (Empty);
  
  // Bob signale la fin de la session
  rpc EndReconciliation (EndRequest) returns (Empty);
}

// Message: Bob demande des parités
message ParityRequest {
  repeated BlockInfo blocks = 1;
}

// Information sur un bloc
message BlockInfo {
  string shuffle_id = 1;     // ID du shuffle utilisé (as string to support large numbers)
  int32 start_index = 2;     // Index de début du bloc
  int32 end_index = 3;       // Index de fin du bloc
}

// Message: Alice répond avec les parités
message ParityResponse {
  repeated int32 parities = 1;  // Liste de 0 ou 1
}

// Message: Bob démarre une reconciliation
message StartRequest {
  string algorithm_name = 1;
}

// Message: Bob termine une reconciliation
message EndRequest {
  string algorithm_name = 1;
}

// Message vide
message Empty {}
